### ISSUE TITLE
impl PoS inflation system

### ISSUE THREAD

**Issue by tzemanovic** [anoma/namada#340]
This is the master set of issues for implementing the PoS inflation and rewards systems.

Designed as specified in:
- #230 
- #283 
- #613 
- #788

Currently, almost all of the code is in #388, but the plan is to separate this down into smaller PRs.

Individual items and issues (checked if completely addressed so far in #388 or in another PR):
- [x] #366 
- [x] #353 
- [x] #455 
- [x] #478 
- [x] #499 
- [x] #436 
- [x] #642 
- [x] #498
- [x] #695 
- [ ] #34 
- [x] #721 
- [x] #787 
- [x] #892
- [ ] #805 
- [ ] #939

Latest details on state of progress:

- Current PoS development is largely based on #740, which changes the PoS code to rely on lazy data structures for storage reading+writing.
- #740 currently needs #913 to be completed in order to properly test the changes. #366 is included in #740.
- #714 is the initial implementation of the inflation and rewards system. This PR includes #353, #499, and #642 from above. Would probably like to get this merged and finish remaining PoS tasks in new PRs on top of this to prevent it from getting too large. Also probably still needs to be changed a bit once rebased onto a complete #740.
- #892 currently relies on #740 and will partially address #787.
- #721 will be mostly developed after #740 and #714 are ready as a base and will address most of #787.
- #34 and #939 will likely be developed once #740 and #714 are ready as a base.

**Comment by brentstone** [comment]
Closing this now that #2006 exists. The only remaining items in here are redelegation and the final PoS rewards.
